org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/InvisibleProjectBuildSupportTest.java

```diff
@@ -313,4 +314,95 @@ public void scheduled(IJobChangeEvent event) {
+			configuration.put(Preferences.JAVA_PROJECT_REFERENCED_LIBRARIES_KEY, new HashMap<String, Object>() {{
+				put("include", include);
+			}});
+			Preferences preferences = Preferences.createFrom(configuration);
+			ReferencedLibraries libraries = new ReferencedLibraries(new HashSet<>(include), new HashSet<>(), new HashMap<>());
+			assertEquals("Configuration with include", libraries, preferences.getReferencedLibraries());
+		}
+
+		{ // Test import of referenced libraries object with include and exclude
+			List<String> include = Arrays.asList("libraries/**/*.jar");
+			List<String> exclude = Arrays.asList("libraries/sources/**");
+			Map<String, Object> configuration = new HashMap<>();
+			configuration.put(Preferences.JAVA_PROJECT_REFERENCED_LIBRARIES_KEY, new HashMap<String, Object>() {{
+				put("include", include);
+				put("exclude", exclude);
+			}});
+			Preferences preferences = Preferences.createFrom(configuration);
+			ReferencedLibraries libraries = new ReferencedLibraries(new HashSet<>(include), new HashSet<>(exclude), new HashMap<>());
+			assertEquals("Configuration with include and exclude", libraries, preferences.getReferencedLibraries());
+		}
+
+		{ // Test import of referenced libraries object with include and sources
+			List<String> include = Arrays.asList("libraries/**/*.jar");
+			Map<String, String> sources = new HashMap<String, String>() {{
+				put("libraries/foo.jar", "libraries/foo-src.jar");
+			}};
+			Map<String, Object> configuration = new HashMap<>();
+			configuration.put(Preferences.JAVA_PROJECT_REFERENCED_LIBRARIES_KEY, new HashMap<String, Object>() {{
+				put("include", include);
+				put("sources", sources);
+			}});
+			Preferences preferences = Preferences.createFrom(configuration);
+			ReferencedLibraries libraries = new ReferencedLibraries(new HashSet<>(include), new HashSet<>(), sources);
+			assertEquals("Configuration with include and sources", libraries, preferences.getReferencedLibraries());
+		}
+
+		{ // Test import of referenced libraries object with include, exclude and sources
+			List<String> include = Arrays.asList("libraries/**/*.jar");
+			List<String> exclude = Arrays.asList("libraries/sources/**");
+			Map<String, String> sources = new HashMap<String, String>() {{
+				put("libraries/foo.jar", "libraries/foo-src.jar");
+			}};
+			Map<String, Object> configuration = new HashMap<>();
+			configuration.put(Preferences.JAVA_PROJECT_REFERENCED_LIBRARIES_KEY, new HashMap<String, Object>() {{
+				put("include", include);
+				put("exclude", exclude);
+				put("sources", sources);
+			}});
+			Preferences preferences = Preferences.createFrom(configuration);
+			ReferencedLibraries libraries = new ReferencedLibraries(new HashSet<>(include), new HashSet<>(exclude), sources);
+			assertEquals("Configuration with include, exclude and sources", libraries, preferences.getReferencedLibraries());
+		}
+
+		{ // Test import of referenced libraries without include (will fall back to default)
```

Comment:
I'd expect exclusions and sources to be taken into account

Reply:
I've just added the cases to test setting with exclude, sources, or exclude + sources.